### PR TITLE
Remove code duplication in segment operations

### PR DESCRIPTION
**Pull request by fritzrehde**
The segment operations all share similar code (looping through the memory pointer to use the ARM `stg` instruction to tag an entire memory region). This `stg` loop functionality has been extracted to a seperate helper function (loop is not a great name though, as we will try to add e.g. the `stg2` instruction in the future which will ommit the need for loops in some cases, hence maybe `stg_region` or similar might be more suitable).
The duplicated code to be removed has been commented out for now (yes, lavaflow anti-pattern, I know), because I am not quite sure my refactored solution is entirely semantically equivalent yet. You will notice that the lines 2593 and 2528 differ in the original code, and the line 2593 is the one I used in the helper function, as I think the `tagged_ptr` is passed unnecessarily to the final basic block (it isn't changed in any of the other blocks). Let me know if my assumption here is incorrect though, and I will fix it.

### CODE REVIEW

**Comment by fritzrehde** [comment]
Also important to note would be that compiling any wasm examples containg the `segment_free` operation (e.g. `segment-free.wast` in the tests mem_safety folder) still fails for some reason.

**Comment by fritzrehde** [comment]
The bug that caused the failing compilation of `segment-free.wast` has been fixed in the `wasm-tools` repo and it compiles now.

File: tests/mem_safety/segment-free.wast [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+        (local.get $ptr)
+        (i32.load_segment)
+
+        (local.get $ptr)
+        (i32.const 16)
+        (segment.free)
+        (return)
+  )
+)
```
Comment (martin-fink) [inline tests/mem_safety/segment-free.wast:21]: You could change the test case to the following:

```suggestion
(module
  (memory 1 1)
  (global $sp (mut i32) (i32.const 64))
  (func (export "foo") (param $idx i64) (result i32)
    (local $ptr i64)
    (i32.const 0)
    (i32.const 16)
    (segment.stack_new)
    (local.tee $ptr)
    (local.get $idx)
    (i64.const 2)
    (i64.shl)
    (i64.add)
    (i32.const 42)
    (i32.store_segment)

    (local.get $ptr)
    (local.get $idx)
    (i64.const 2)
    (i64.shl)
    (i64.add)
    (i32.load_segment)

    (local.get $ptr)
    (i32.const 16)
    (segment.free)
    (return)
  )
)
```

This basically just adds a parameter, so we can access the memory with a user-supplied index and check if memory protection works in the following way:

```shell
$ ./wasmtime run --allow-precompiled --cranelift-enable use_mte --wasm-features=mem-safety segment-free.cwasm --invoke foo 3
42
$ ./wasmtime run --allow-precompiled --cranelift-enable use_mte --wasm-features=mem-safety segment-free.cwasm --invoke foo 4
Caused by:
    0: failed to invoke `foo`
    1: error while executing at wasm backtrace:
           0:   0x45 - <unknown>!<wasm function 0>
    2: memory fault at wasm address 0x10 in linear memory of size 0x10000
    3: wasm trap: out of bounds memory access
```

File: cranelift/wasm/src/code_translator.rs [hunk 2]
```diff
@@ -2479,155 +2479,62 @@ pub fn translate_operator<FE: FuncEnvironment + ?Sized>(
...
-            let tagged_pointer = builder.ins().arm64_irg(base);
-
-            let inner_block = block_with_params(builder, [ValType::I32, ValType::I64], environ)?;
-            let next_inner_block =
-                block_with_params(builder, [ValType::I32, ValType::I64], environ)?;
-            let next_block = block_with_params(builder, [ValType::I64], environ)?;
-
-            builder.ins().jump(inner_block, &[size, tagged_pointer]);
+            // remove existing tag in base_ptr
+            let base_ptr = builder.ins().band_imm(base_ptr, 0x0FFF_FFFF_FFFF_FFFF);
```
Comment (martin-fink) [inline cranelift/wasm/src/code_translator.rs:2499]: MTE tag is bits 56-59 (sorry -- my fault!).

```suggestion
            let base_ptr = builder.ins().band_imm(base_ptr, 0xF0FF_FFFF_FFFF_FFFF);
```

File: cranelift/wasm/src/code_translator.rs [hunk 3]
```diff
@@ -3795,3 +3702,109 @@ fn bitcast_wasm_params<FE: FuncEnvironment + ?Sized>(
...
+    // +---
+
+    assert_16_byte_aligned(size, builder);
+
+    // === Block definitions
+    // inner_block(size, iter_ptr)
+    let inner_block = block_with_params(builder, [ValType::I32, ValType::I64], environ)?;
+
+    // next_inner_block(size, iter_ptr)
+    let next_inner_block = block_with_params(builder, [ValType::I32, ValType::I64], environ)?;
```
Comment (martin-fink) [inline cranelift/wasm/src/code_translator.rs:3759]: And this to `loop_body_block`

**Review by martin-fink — COMMENTED** [review]
I've tested the code for `segment.free` and it works! Looks good so far 🙂

**Comment by fritzrehde** [comment]
I think I implemented all the changes you suggested.

**Comment by martin-fink** [comment]
Nice! If you squash your related commits and mark this PR as ready, we can merge it.

**Review by martin-fink — APPROVED** [review]

**Pull request merged**